### ISSUE TITLE
FileStorage occasionally get corrupted. No way to recover

### ISSUE THREAD

**Issue by LarsLiden** [Microsoft/botbuilder-js#104]
We've found using FileStorage that occasionally the files on disk get corrupted (i.e. file is truncated or otherwise not valid JSON).  Not quite sure what causes this condition. Perhaps when node process is prematurely stopped?

More importantly, there doesn't appear to be a graceful way to recover from this condition other than manually deleting the storage files from disk.   One can catch the failed reads, but there's no obvious way to clear the bad file.

Any suggestions?

**Comment by cleemullins** [comment]
Bigger picture, the Memory + File solutions we have now are just not rich enough to build a real bot atop. If nothing else, you can't build a scalable bot using either as state is affinitized to the local machine. The best choice - really the only choice - is to move to a more mature state solution. CosmosDB is where we anticipate most people storing their state. 

We could certainly expose a hook in the FileStorage system that allows your code to decide what to do if things get corrupted.  This could even be a state flag you pass in. It would certainly also be good if the storage provider didn't corrupt itself...

**Comment by LarsLiden** [comment]
Completely agree.  I created a Redis middleware plugin (which we show in one of our sample apps), but we need a quick way for users to get off the ground w/o having to create a storage account, provide keys, etc.   We default to FileStorage for that case.

Would be awesome such a flag were added so we can recover from corrupted data.  Currently our users get stuck and we have to tell them to manually delete the files

**Comment by Stevenic** [comment]
I can take a look at fixing this when I port the FileStorage provider over to the new M2 branch.  I've done just about everything but that actually.

**Comment by LarsLiden** [comment]
Thanks!

**Comment by LarsLiden** [comment]
Thanks, Steve.  So the subsequent write will replace the corrupted file?

**Comment by Stevenic** [comment]
yes... Now if a read error occurs we log it as a warning but then initialize the state to be empty. The next save should overwrite the corrupted file.
